python/tests/test_types.py

```diff
@@ -204,3 +206,388 @@ def test_non_parameterized_type_equality(input_index, input_type, check_index, c
+
+
+def test_repr_decimal():
+    assert repr(DecimalType(19, 25)) == "DecimalType(precision=19, scale=25)"
+
+
+def test_serialization_nestedfield():
+    expected = '{"id": 1, "name": "required_field", "type": "string", "required": true, "doc": "this is a doc"}'
+    actual = NestedField(1, "required_field", StringType(), True, "this is a doc").json()
+    assert expected == actual
+
+
+def test_serialization_nestedfield_no_doc():
+    expected = '{"id": 1, "name": "required_field", "type": "string", "required": true}'
+    actual = NestedField(1, "required_field", StringType(), True).json()
+    assert expected == actual
+
+
+def test_str_nestedfield():
+    assert str(NestedField(1, "required_field", StringType(), True)) == "1: required_field: required string"
+
+
+def test_repr_nestedfield():
+    assert (
+        repr(NestedField(1, "required_field", StringType(), True))
+        == "NestedField(field_id=1, name='required_field', field_type=StringType(), required=True)"
+    )
+
+
+def test_nestedfield_by_alias():
+    # We should be able to initialize a NestedField by alias
+    expected = NestedField(1, "required_field", StringType(), True, "this is a doc")
+    actual = NestedField(**{"id": 1, "name": "required_field", "type": "string", "required": True, "doc": "this is a doc"})
+    assert expected == actual
+
+
+def test_deserialization_nestedfield():
+    expected = NestedField(1, "required_field", StringType(), True, "this is a doc")
+    actual = NestedField.parse_raw(
+        '{"id": 1, "name": "required_field", "type": "string", "required": true, "doc": "this is a doc"}'
+    )
+    assert expected == actual
+
+
+def test_deserialization_nestedfield_inner():
+    expected = NestedField(1, "required_field", StringType(), True, "this is a doc")
+    actual = TestType.parse_raw('{"id": 1, "name": "required_field", "type": "string", "required": true, "doc": "this is a doc"}')
+    assert expected == actual.__root__
+
+
+def test_serialization_struct():
+    actual = StructType(
+        NestedField(1, "required_field", StringType(), True, "this is a doc"), NestedField(2, "optional_field", IntegerType())
+    ).json()
+    expected = (
+        '{"type": "struct", "fields": ['
+        '{"id": 1, "name": "required_field", "type": "string", "required": true, "doc": "this is a doc"}, '
+        '{"id": 2, "name": "optional_field", "type": "int", "required": true}'
+        "]}"
+    )
+    assert actual == expected
+
+
+def test_deserialization_struct():
+    actual = StructType.parse_raw(
+        """
+    {
+        "type": "struct",
+        "fields": [{
+                "id": 1,
+                "name": "required_field",
+                "type": "string",
+                "required": true,
+                "doc": "this is a doc"
+            },
+            {
+                "id": 2,
+                "name": "optional_field",
+                "type": "int",
+                "required": true,
+                "doc": null
```

Comment:
Minor: Is it possible to omit the field rather than making it explicitly null?

Reply:
This is a deserialization input, so that should be possible. I've updated the code.